File: ad-sdk/src/main/java/plugin/adsdk/service/BaseActivity.java
package plugin.adsdk.service;

import static android.os.Build.VERSION.SDK_INT;
import static plugin.adsdk.service.AdsUtility.config;

import android.app.Dialog;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.OnUserEarnedRewardListener;
import com.google.android.gms.ads.rewarded.RewardItem;
import com.google.android.gms.ads.rewarded.RewardedAd;
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.CompositeMultiplePermissionsListener;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;

import java.util.Date;
import java.util.List;

import plugin.adsdk.R;
import plugin.adsdk.extras.BaseLauncherActivity;
import plugin.adsdk.extras.NetworkChangeReceiver;
import plugin.adsdk.extras.ThankYouBottomSheet;
import plugin.adsdk.service.utils.ValidationHandler;

@SuppressWarnings("unused")
//@Obfuscate
public class BaseActivity extends AppCompatActivity {
    public static final String PARAM_FOR_RESULT = "FOR_RESULT";
    public static final String PARAM_REQUEST_CODE = "REQUEST_CODE";
    public static final int REQUEST_CODE_PERMISSION = 3256;
    public static final int REQUEST_CODE_APP_UPDATE = 7588;

    private static Dialog progressDialog;
    private NetworkChangeReceiver receiver = null;

    public static void dismissDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

    public void showDialog() {
        if (progressDialog != null && !progressDialog.isShowing() && !isFinishing()) {
            progressDialog.show();
        }
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    protected void setupNetworkChangeReceiver() {
        receiver = new NetworkChangeReceiver(this::networkStateChanged);
        IntentFilter filter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
        registerReceiver(receiver, filter);
    }

    @Override
    protected void onResume() {
        super.onResume();
//        if (!(this instanceof BaseLauncherActivity) && ValidationHandler.hasDebuggingEnabled(this))
//            ValidationHandler.showDeveloperDialog(this, null);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        try {
            unregisterReceiver(receiver);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected void networkStateChanged(NetworkChangeReceiver.NetworkState state) {
    }

    protected void bannerAd() {
        bannerAd(findViewById(R.id.banner_ad_container));
    }

    public void bannerAd(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.bannerAd)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestBannerAd(this, viewGroup);
    }

    public void bannerAdLanguage(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.bannerAdLanguage)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestBannerAdLanguage(this, viewGroup);
    }
    protected void mediumRectangleBannerAd() {
        mediumRectangleBannerAd(findViewById(R.id.banner_ad_container));
    }

    public void mediumRectangleBannerAd(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.bannerAd)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestMediumRectangleBanner(this, viewGroup);
    }

    protected void nativeAdMedium() {
        nativeAdMedium(findViewById(R.id.native_ad_container));
    }

    public void nativeAdMedium(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.nativeAd)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestNativeAdMedium(this, viewGroup);
    }

    protected void nativeAdSmall() {
        nativeAdSmall(findViewById(R.id.native_ad_container));
    }

    public void nativeAdSmall(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.nativeAd)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestNativeAdSmall(this, viewGroup, R.layout.ad_native_small);
    }

    protected void nativeAd() {
        nativeAd(findViewById(R.id.native_ad_container));
    }

    public void nativeAd(ViewGroup viewGroup) {
        if (TextUtils.isEmpty(config.adMob.nativeAd)) {
            viewGroup.setVisibility(View.GONE);
            return;
        }
        if (AdsUtility.optimizeAppOpen()) {
            AppOpenManager.blockAppOpen(this); //limiting app-open
        }
        AdsUtility.requestNativeAd(this, viewGroup);
    }

    public void showInterstitial(final Intent intent) {
        showInterstitial(() -> startActivity(intent));
    }

    public void showInterstitial(final BaseCallback callback) {
//        if (config.allowAdIntervalInterstitial) {
//            //interstitial for time based interval
//            long now = new Date().getTime();
//            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(this);
//            long lastAdTime = pref.getLong("last_ad_time_interstitial", now);
//            long allowedDelta = config.adIntervalTimeInterstitial * 1000L;
//            long currentAdTimeDelta = now - lastAdTime;
//            if (currentAdTimeDelta >= allowedDelta) {
//                pref.edit().putLong("last_ad_time_interstitial", now).apply();
//                AdsUtility.showInterstitial(this, callback);
//            } else {
//                callback.completed();
//            }
//            return;
//        }
        if (AdsUtility.currentActivityCount == 0) {
            if (config.replaceInterWithAppOpen) {
                //create a function which complies with on-demand as well as preload
                AppOpenManager.get().showAppOpen(callback);
            } else {
                //check for alternative inter & app-open
                if (config.alternateAds) {
                    if (AdsUtility.shouldAlternate) {
                        AdsUtility.showInterstitial(this, callback);
                        AdsUtility.shouldAlternate = false;
                    } else {
                        AppOpenManager.get().showAppOpen(callback);
                        AdsUtility.shouldAlternate = true;
                    }
                } else {
                    AdsUtility.showInterstitial(this, callback);
                }
            }
        } else {
//            AdsUtility.currentActivityCount = (AdsUtility.currentActivityCount + 1) % config.activityCount;
            callback.completed();
        }
    }

    protected void showInitialInterstitial(final BaseCallback callback) {
        AdsUtility.showInterstitial(this, callback);
    }

    //Source Needs Update
    private RewardItem rewardItem = null;

    public void showAdmobRewardedVideoAd(OnUserEarnedRewardListener callback) {
        if (TextUtils.isEmpty(config.adMob.videoAd)) {
            callback.onUserEarnedReward(rewardItem);
            return;
        }

        setupDialog();
        progressDialog.show();

        final FullScreenContentCallback fullScreenContentCallback = new FullScreenContentCallback() {
            @Override
            public void onAdDismissedFullScreenContent() {
                super.onAdDismissedFullScreenContent();
                callback.onUserEarnedReward(rewardItem);
                rewardItem = null;
            }
        };

        OnUserEarnedRewardListener rewardCallback = reward -> rewardItem = reward;

        RewardedAd.load(
                this,
                config.adMob.videoAd,
                new AdRequest.Builder().build(),
                new RewardedAdLoadCallback() {
                    @Override
                    public void onAdFailedToLoad(@NonNull LoadAdError loadAdError) {
                        dismissDialog();
                        Toast.makeText(BaseActivity.this, "Failed to fetch reward!", Toast.LENGTH_SHORT).show();
                        callback.onUserEarnedReward(rewardItem);
                    }

                    @Override
                    public void onAdLoaded(@NonNull RewardedAd rewardedAd) {
                        rewardedAd.setFullScreenContentCallback(fullScreenContentCallback);
                        rewardedAd.show(BaseActivity.this, rewardCallback);
                        dismissDialog();
                    }
                });
    }

    public void startActivity(Intent intent) {
        if (intent != null) {
            if (intent.getBooleanExtra(PARAM_FOR_RESULT, false)) {
                //noinspection deprecation
                startActivityForResult(intent, intent.getIntExtra(PARAM_REQUEST_CODE, -1));
            } else {
                super.startActivity(intent);
            }
        }
    }

    void setupDialog() {
        progressDialog = new Dialog(this);
        progressDialog.setContentView(R.layout.ad_loading);
        progressDialog.setCancelable(false);
        progressDialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
    }

    public void statusBarDark() {
        if (SDK_INT >= Build.VERSION_CODES.M) {
            int flags = getWindow().getDecorView().getSystemUiVisibility(); // get current flag
            flags = flags ^ View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR; // use XOR here for remove LIGHT_STATUS_BAR from flags
            getWindow().getDecorView().setSystemUiVisibility(flags);
        }
    }

    //use REQUEST_CODE_PERMISSION in onActivityResult() method to re-verify permission approval
    public void checkRunTimePermission(final BaseCallback callback, Boolean showAd, String... permissionArrays) {
        if (SDK_INT >= Build.VERSION_CODES.M) {
            if (hasPermissions(permissionArrays)) {
                if (showAd) showInterstitial(callback);
                else callback.completed();
            } else {
                Dexter.withContext(this)
                        .withPermissions(permissionArrays)
                        .withListener(new CompositeMultiplePermissionsListener(new MultiplePermissionsListener() {
                            @Override
                            public void onPermissionsChecked(MultiplePermissionsReport multiplePermissionsReport) {
                                if (multiplePermissionsReport.areAllPermissionsGranted()) {
                                    if (showAd) showInterstitial(callback);
                                    else callback.completed();
                                } else {
                                    showPermissionSnack();
                                }
                            }

                            @Override
                            public void onPermissionRationaleShouldBeShown(List<PermissionRequest> list, PermissionToken permissionToken) {
                                permissionToken.continuePermissionRequest();
                            }
                        })).onSameThread().check();
            }
        } else {
            if (showAd) showInterstitial(callback);
            else callback.completed();
        }
    }

    private void showPermissionSnack() {
        String message = "All those permissions are needed for working all functionality";
        Snackbar snackbar = Snackbar.make(findViewById(android.R.id.content), message, BaseTransientBottomBar.LENGTH_INDEFINITE)
                .setAction("Settings", view -> openSettingsDialog());
        snackbar.setActionTextColor(Color.RED);

        View sbView = snackbar.getView();
        TextView textView = sbView.findViewById(com.google.android.material.R.id.snackbar_text);
        textView.setTextColor(Color.YELLOW);
        snackbar.show();
    }

    private void openSettingsDialog() {
        final String EXTRA_FRAGMENT_ARG_KEY = ":settings:fragment_args_key";
        final String EXTRA_SHOW_FRAGMENT_ARGUMENTS = ":settings:show_fragment_args";
        final String EXTRA_SYSTEM_ALERT_WINDOW = "permission_settings";

        Bundle bundle = new Bundle();
        bundle.putString(EXTRA_FRAGMENT_ARG_KEY, EXTRA_SYSTEM_ALERT_WINDOW);

        Uri uri = Uri.fromParts("package", getPackageName(), null);
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                .setData(uri)
                .putExtra(EXTRA_FRAGMENT_ARG_KEY, EXTRA_SYSTEM_ALERT_WINDOW)
                .putExtra(EXTRA_SHOW_FRAGMENT_ARGUMENTS, bundle);
        //noinspection deprecation
        startActivityForResult(intent, REQUEST_CODE_PERMISSION);
    }

    public boolean hasPermissions(String... permissions) {
        if (permissions != null) {
            for (String permission : permissions) {
                if (ActivityCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                    return false;
                }
            }
        }
        return true;
    }

    public void showThankYouSheet() {
        ThankYouBottomSheet addPhotoBottomDialogFragment =
                ThankYouBottomSheet.newInstance();
        addPhotoBottomDialogFragment.show(getSupportFragmentManager(),
                ThankYouBottomSheet.TAG);
    }

    protected long mLastClickTime = 0;

    protected void backPressed(BaseCallback callback) {
        if (SystemClock.elapsedRealtime() - mLastClickTime < 1000) {
            return;
        }
        mLastClickTime = SystemClock.elapsedRealtime();
        if (AdsUtility.config.adOnBack) {
            showInterstitial(callback);
        } else {
            callback.completed();
        }
    }

    protected void backPressed() {
        backPressed(super::onBackPressed);
    }

    protected void exitBackPressed(BaseCallback callback) {
        if (AdsUtility.startScreenCount != 0) {
            backPressed();
        } else {
            callback.completed();
        }
    }

    protected void exitBackPressed() {
        exitBackPressed(this::showThankYouSheet);
    }

    protected ActivityResultLauncher<Intent> appOpenBlockLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> AppOpenManager.overrideAppOpenShow(false)
    );
}

File: app/src/main/java/calculator/lock/hide/photo/video/calculatorvault/safe/photovault/hideru/vault/ui/base/BaseActivity.kt
package calculator.lock.hide.photo.video.calculatorvault.safe.photovault.hideru.vault.ui.base

import android.Manifest
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Build.VERSION.SDK_INT
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.view.WindowInsetsCompat
import androidx.viewbinding.ViewBinding
import calculator.lock.hide.photo.video.calculatorvault.safe.photovault.hideru.vault.R
import calculator.lock.hide.photo.video.calculatorvault.safe.photovault.hideru.vault.adsModule.AdsUtils
import calculator.lock.hide.photo.video.calculatorvault.safe.photovault.hideru.vault.adsModule.setOnClickListener
import calculator.lock.hide.photo.video.calculatorvault.safe.photovault.hideru.vault.utils.MyPreferences
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.karumi.dexter.Dexter
import com.karumi.dexter.MultiplePermissionsReport
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.multi.CompositeMultiplePermissionsListener
import com.karumi.dexter.listener.multi.MultiplePermissionsListener
import com.simplemobiletools.commons.extensions.addBit
import com.simplemobiletools.commons.extensions.isUsingGestureNavigation
import com.simplemobiletools.commons.extensions.navigationBarHeight
import com.simplemobiletools.commons.extensions.onApplyWindowInsets
import com.simplemobiletools.commons.extensions.removeBit
import com.zeugmasolutions.localehelper.LocaleHelper
import com.zeugmasolutions.localehelper.LocaleHelperActivityDelegate
import com.zeugmasolutions.localehelper.LocaleHelperActivityDelegateImpl
import plugin.adsdk.service.AdsUtility
import plugin.adsdk.service.BaseActivity
import java.util.Locale

abstract class BaseActivity<B : ViewBinding> : BaseActivity() {

    lateinit var binding: B
    lateinit var myPreferences: MyPreferences
    abstract fun getViewBinding(): B

    abstract fun initData()

    override fun onCreate(savedInstanceState: Bundle?) {
        val decorView = window.decorView
        var newUiOptions = decorView.systemUiVisibility
        newUiOptions = newUiOptions or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        newUiOptions = newUiOptions or View.SYSTEM_UI_FLAG_IMMERSIVE
        newUiOptions = newUiOptions or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        decorView.systemUiVisibility = newUiOptions
        super.onCreate(savedInstanceState)
        myPreferences = MyPreferences.getPreferences(this)!!
        localeDelegate.onCreate(this)
        binding = getViewBinding()
        setContentView(binding.root)
        initData()
        setupNetworkChangeReceiver()
    }

    private fun handleNavigationAndScrolling() {
        if (navigationBarHeight > 0 || isUsingGestureNavigation()) {
            window.decorView.systemUiVisibility =
                window.decorView.systemUiVisibility.addBit(View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
            // Don't touch this. Window Inset API often has a domino effect and things will most likely break.
            onApplyWindowInsets {
                val insets =
                    it.getInsets(WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.ime())
            }
        } else {
            window.decorView.systemUiVisibility =
                window.decorView.systemUiVisibility.removeBit(View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION)
        }
    }


    fun showExitDialog() {
        val dialog = Dialog(this)
        dialog.setContentView(R.layout.dialog_exit)
        val width = (resources.displayMetrics.widthPixels * 0.85).toInt()
        dialog.window?.setBackgroundDrawableResource(R.drawable.bg_card_edit)
        dialog.window?.setLayout(width, ViewGroup.LayoutParams.WRAP_CONTENT)
        dialog.findViewById<TextView>(R.id.tvShare).isSelected = true
        dialog.findViewById<TextView>(R.id.tvQuit).isSelected = true
        dialog.findViewById<TextView>(R.id.tvExitTextTitle).isSelected = true
        dialog.findViewById<TextView>(R.id.tvExitTextTitle1).isSelected = true

        dialog.findViewById<TextView>(R.id.tvShare).setOnClickListener(500L) {
            dialog.dismiss()
            AdsUtility.rateUs(this)
        }
        dialog.findViewById<TextView>(R.id.tvQuit).setOnClickListener(500L) {
            dialog.dismiss()
            AdsUtils.logActivityOpen(this, "MainScreen", "AppExited")
//            exitProcess(0)
            finishAffinity()
        }
        dialog.show()
    }

    var appOpenBlockLauncher: ActivityResultLauncher<Intent> = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {}

    val REQUEST_CODE_PERMISSION = 3256

    open fun permissionsNotification(): Array<String>? {
        val p: Array<String> = if (SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permission33Notification
        } else {
            arrayOf()
        }
        return p
    }

    fun permissionsStorage(): Array<String> {
        val p: Array<String> = if (SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permission33
        } else {
            permissionAll
        }
        return p
    }

    private var permissionQ = arrayOf<String>(
        Manifest.permission.MANAGE_EXTERNAL_STORAGE
    )

    private var permissionAll = arrayOf<String>(
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_EXTERNAL_STORAGE
    )

    @RequiresApi(api = Build.VERSION_CODES.TIRAMISU)
    private var permission33 = arrayOf<String>(
        Manifest.permission.READ_MEDIA_IMAGES,
        Manifest.permission.READ_MEDIA_VIDEO,
        Manifest.permission.POST_NOTIFICATIONS
    )

    @RequiresApi(api = Build.VERSION_CODES.TIRAMISU)
    private var permission33Notification = arrayOf<String>(
        Manifest.permission.POST_NOTIFICATIONS
    )

    open fun checkRunTimePermission(
        vararg permissionArrays: String?,
        listener: () -> Unit
    ) {
        if (SDK_INT >= Build.VERSION_CODES.M) {
            if (hasPermissions(*permissionArrays)) {
                listener.invoke()
            } else {
                Dexter.withContext(this).withPermissions(*permissionArrays)
                    .withListener(CompositeMultiplePermissionsListener(object :
                        MultiplePermissionsListener {
                        override fun onPermissionsChecked(multiplePermissionsReport: MultiplePermissionsReport) {
                            if (multiplePermissionsReport.areAllPermissionsGranted()) {
                                listener.invoke()
                            } else {
                                showPermissionSnack()
                            }
                        }

                        override fun onPermissionRationaleShouldBeShown(
                            list: MutableList<com.karumi.dexter.listener.PermissionRequest>?,
                            permissionToken: PermissionToken?
                        ) {
                            permissionToken?.continuePermissionRequest()
                        }

                    })).onSameThread().check()
            }
        } else {

        }
    }

    override fun hasPermissions(vararg permissions: String?): Boolean {
        if (permissions != null) {
            for (permission in permissions) {
                if (ActivityCompat.checkSelfPermission(
                        this, permission!!
                    ) != PackageManager.PERMISSION_GRANTED
                ) {
                    return false
                }
            }
        }
        return true
    }

    open fun showPermissionSnack() {
        val message = "All those permissions are needed for working all functionality"
        val snackbar = Snackbar.make(
            findViewById<View>(android.R.id.content),
            message,
            BaseTransientBottomBar.LENGTH_INDEFINITE
        ).setAction("Settings") { _: View? -> openSettingsDialog() }
        snackbar.setActionTextColor(Color.RED)
        val sbView = snackbar.view
        val textView = sbView.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
        textView.setTextColor(Color.YELLOW)
        snackbar.show()
    }

    open fun openSettingsDialog() {
        val EXTRA_FRAGMENT_ARG_KEY = ":settings:fragment_args_key"
        val EXTRA_SHOW_FRAGMENT_ARGUMENTS = ":settings:show_fragment_args"
        val EXTRA_SYSTEM_ALERT_WINDOW = "permission_settings"
        val bundle = Bundle()
        bundle.putString(EXTRA_FRAGMENT_ARG_KEY, EXTRA_SYSTEM_ALERT_WINDOW)
        val uri = Uri.fromParts("package", packageName, null)
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).setData(uri)
            .putExtra(EXTRA_FRAGMENT_ARG_KEY, EXTRA_SYSTEM_ALERT_WINDOW)
            .putExtra(EXTRA_SHOW_FRAGMENT_ARGUMENTS, bundle)
        startActivityForResult(intent, REQUEST_CODE_PERMISSION)
    }

    val localeDelegate: LocaleHelperActivityDelegate = LocaleHelperActivityDelegateImpl()

    override fun getDelegate() = localeDelegate.getAppCompatDelegate(super.getDelegate())

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(localeDelegate.attachBaseContext(newBase))
    }

    override fun onResume() {
        super.onResume()
//        setThemeNN()
        localeDelegate.onResumed(this)
    }

    override fun onPause() {
        super.onPause()
        localeDelegate.onPaused()
    }

    override fun createConfigurationContext(overrideConfiguration: Configuration): Context {
        val context = super.createConfigurationContext(overrideConfiguration)
        return LocaleHelper.onAttach(context)
    }

    override fun getApplicationContext(): Context =
        localeDelegate.getApplicationContext(super.getApplicationContext())

    open fun updateLocale(locale: Locale) {
        localeDelegate.setLocale(this, locale)
    }


}